src/rdkafka.c

```diff
@@ -270,7 +279,463 @@ void rd_kafka_log0 (const rd_kafka_conf_t *conf,
+			    "Must supply non-null token principal name metadata");
+                return -1;
+        }
+        char errstr[512];
+        if (check_oauthbearer_extension_value(token_value, errstr, sizeof(errstr)) == -1) {
+                rd_kafka_set_fatal_error(rk, RD_KAFKA_RESP_ERR__INVALID_ARG, "%s", errstr);
+                return -1;
+        }
+        size_t i;
+        for (i = 0; i + 1 < extension_size; i += 2) {
+                if (check_oauthbearer_extension_key(extensions[i], errstr, sizeof(errstr)) == -1 ||
+                        check_oauthbearer_extension_value(extensions[i + 1], errstr, sizeof(errstr)) == -1) {
+                                rd_kafka_set_fatal_error(rk,
+                                        RD_KAFKA_RESP_ERR__INVALID_ARG, "%s", errstr);
+                                return -1;
+                }
+        }
+        rd_ts_t now_ms = rd_clock() / 1000;
+        if (md_lifetime_ms <= now_ms) {
+		rd_kafka_set_fatal_error(rk, RD_KAFKA_RESP_ERR__INVALID_ARG,
+			    "Must supply an unexpired token: now=%lli ms, exp=%lli ms",
+                            now_ms, md_lifetime_ms);
+                return -1;
+        }
+        rwlock_wrlock(&rk->rk_oauthbearer->refresh_lock);
+        // rd_free() ends up as a no-op only on initial invocation
+        rd_free(rk->rk_oauthbearer->md_principal_name);
+        rk->rk_oauthbearer->md_principal_name = rd_strdup(md_principal_name);
+        rd_free(rk->rk_oauthbearer->token_value);
+        rk->rk_oauthbearer->token_value = rd_strdup(token_value);
+        rk->rk_oauthbearer->md_lifetime_ms = md_lifetime_ms;
+        // Schedule a refresh 80% through its remaining lifetime
+        rk->rk_oauthbearer->refresh_after_ms = now_ms + 0.8 * (md_lifetime_ms - now_ms);
+        rd_list_destroy(&rk->rk_oauthbearer->extensions);
+        for (i = 0; i + 1 < extension_size; i += 2) {
+                rd_list_add(&rk->rk_oauthbearer->extensions,
+                        rd_strtup_new(extensions[i], extensions[i + 1]));
+        }
+        rk->rk_oauthbearer->errstr[0] = '\0';
+        rwlock_wrunlock(&rk->rk_oauthbearer->refresh_lock);
+        rd_kafka_dbg(rk, SECURITY, "BRKMAIN",
+                "Waking up waiting brokers after setting token");
+        rd_kafka_all_brokers_wakeup(rk, RD_KAFKA_BROKER_STATE_TRY_CONNECT);
+
+        return 0;
+}
+
+void rd_kafka_oauthbearer_set_token_failure(rd_kafka_t *rk,
+                const char *errstr) {
+        rwlock_wrlock(&rk->rk_oauthbearer->refresh_lock);
+        strncpy(rk->rk_oauthbearer->errstr, errstr,
+                sizeof(rk->rk_oauthbearer->errstr));
+        /* Leave any existing token because it may have some life left */
+        /* Schedule a refresh for 10 seconds later */
+        rk->rk_oauthbearer->refresh_after_ms = rd_clock() / 1000 + 10 * 1000;
+        rwlock_wrunlock(&rk->rk_oauthbearer->refresh_lock);
+}
+
+static void parse_unsecured_jws_config_value_for_prefix(char **loc, const char *prefix,
```

Comment:
Have it return -1 on error, or 0 on success.

Reply:
👍